service_linux.go

```diff
@@ -395,23 +416,43 @@ func programIngress(gwIP net.IP, ingressPorts []*PortConfig, isDelete bool) erro
 			}
 		}
 
+		defer func() {
+			if portErr != nil && !isDelete {
+				rule := strings.Fields(fmt.Sprintf("-t nat %s %s -p %s --dport %d -j DNAT --to-destination %s:%d", rollbackAddDelOpt,
+					ingressChain, strings.ToLower(PortConfig_Protocol_name[int32(iPort.Protocol)]), iPort.PublishedPort, gwIP, iPort.PublishedPort))
```

Comment:
Hrm...  Rather than stacking up a defer function per programmed port, I'd consider having a list of "rollback rules".  Then, there could just be one defer function defined before the loop that looks something like this:
```
var portErr error
rollbackRules := make([][]string, 0, len(ingressPorts)*3)
defer func() {
        if portErr != nil && !isDelete {
                for _, rule := range rollbackRules {
                        if err := iptables.RawCombinedOutput(rule...); err != nil {
                                logrus.Warnf("setting up rule failed while rolling back:, %v: %v", rule, err)
                        }
                }
        }
}
```
Then at this point in the code, you don't need this to be in a defer block, you just do:
```
rollBackRule := strings.Fields(...)
rollbackRules = append(rollbackRules, rollbackRule)
```

Incidentally, moving portErr outside of the loop has the added benefit that it can help recover from proxy errors: see below.